lib/sshutils/sftp/sftp.go

```diff
@@ -0,0 +1,362 @@
+/*
+Copyright 2022 Gravitational, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// Package sftp handles file transfers client-side via SFTP
+package sftp
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"os"
+	"path"
+	"time"
+
+	"github.com/gravitational/teleport"
+	"github.com/gravitational/teleport/lib/sshutils/scp"
+	"github.com/gravitational/trace"
+
+	"github.com/pkg/sftp"
+	"github.com/schollz/progressbar/v3"
+	log "github.com/sirupsen/logrus"
+	"golang.org/x/crypto/ssh"
+)
+
+// Options control aspects of a file transfer
+type Options struct {
+	// Recursive indicates recursive file transfer
+	Recursive bool
+	// PreserveAttrs preserves access and modification times
+	// from the original file
+	PreserveAttrs bool
+}
+
+// Config describes the settings of a file transfer
+type Config struct {
+	srcPaths []string
+	dstPath  string
+	srcFS    FileSystem
+	dstFS    FileSystem
+	opts     Options
+
+	// ProgressWriter is a writer for printing the progress
+	// (used only on the client)
+	ProgressWriter io.Writer
+	// Log optionally specifies the logger
+	Log log.FieldLogger
+}
+
+// FileSystem describes file operations to be done either locally or over SFTP
+type FileSystem interface {
+	// SetContext sets the context for the filesystem used for cancellation
+	SetContext(ctx context.Context)
```

Comment:
Can you please remove that method and pass context as a parameter to each method that needs it? 
https://github.com/golang/go/wiki/CodeReviewComments#contexts

Reply:
As discussed on Slack, I think the interface is valid given it's use case, but changing it as you suggest will make it more idiomatic and flexible, so I agree it does need to be changed. Done